keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,750 @@
+   ```
+   or alternatively by:
+   ```yaml
+   x-kubernetes-int-or-string: true
+   allOf:
+   - anyOf: // optionally provided and accepted
+     - type: integer
+     - type: string
+   - pattern: abc
+     anyOf:
+     - minimum: 42
+       maximum: 50
+     - minimum: 52
+       maximum: 60
+   ```
+   The later is what is also published.
+   
+   After dropping the value validation, we get:
+   ```yaml
+   x-kubernetes-int-or-string: true
+   ```   
+
+## Design Details
+
+### Scope of necessary Changes
+
+* To implement this KEP we won't have to modify CRD validation in any way. We can keep using the go-openapi based validation algorithm.
+* To implement this KEP we won't have to modify kubectl in any way. We can though add support for the new `x-kubernetes-*` fields, though we will probably wait for OpenAPI v3 publishing and support in kubectl before doing that.
+* We don't have to change the API types as described above in the proposal, but we can implement the structural schema check via a number of additional validations on-top of the existing API types.
+
+### API Compatibility Plan
+
+* When creating `v1` CRDs, only structural schemas (with value-validation) are allowed.
+Updating a schema in a `v1` CRD to a non-structural schema is only allowed if the original schema was non-structural already.
+* Non-structural CRDs only allow `v1` schema updates if a structural schema is set.
+* Any new feature and any feature not promoted to beta yet, will be available for structural schemas only:
+  * webhook conversion:
+    * rejected at CRD validation time on create and on update which sets the webhook; user gets validation error
+    * pre-existing webhooks on update will stay allowed
+  * publishing
+    * not published at publish time, `NonStructuralSchema condition is added to CRD status with a warning
+` * pruning (which will be with explicit opt-in in `v1beta1`)
+    * rejected at CRD validation time if pruning is enabled in the CRD; user gets validation error
+  * defaulting
+    * rejected at CRD validation time if defaults are set in the CRD; user gets validation error
+  * `x-kubernetes-union-*` fields:
+    * rejected at CRD validation time if those fields are set; user gets validation error
+
+  This should create natural motivation for the community to phase out old, non-structural schemas without breaking beta compatibility guarantees.
+  
+We add a condition `NonStructuralSchema` to the CRD objects to reflect that information to the user about a non-structural schema. This condition is only created in case its value is `True`. It is removed if it turns to `False` due to schema updates.
+ 
+### Implementation Plan
+
+**For Kubernetes 1.15, must haves:**
+
+* Implement structural-schema API validation for 1.15, but don't reject non-structural objects, only set the NonStructuralSchema condition.
+* Do not publish a CRD schema if NonStructuralSchema condition is true (pre-requisite of promoting publishing to beta).
```

Comment:
even some of the structural schemas cannot be published in openapi v2, due to kubectl handling of additionalProperties, right?

Reply:
as we enforce types and properties outside of the logical junctors, we enforce that more of the spec is publishing (junctors are dropped today). So yes, not all goes through, but much more than before.